### PR TITLE
CURATOR-408 Handle graceful close of ZookKeeper client waiting for all resources to be released

### PR DESCRIPTION
**Pull request by eolivelli**
Add a new close(int timeout) API in order to leverage ZooKeeper#close(timeout) API introduced in ZK 4.5.4-BETA.

Summary of changes:
- bump ZK to 4.5.4-BETA
- add the implementation of the feature
- add a missing license header

### CODE REVIEW

File: curator-client/src/main/java/org/apache/curator/ConnectionState.java [hunk 1]
```diff
@@ -112,14 +112,17 @@ void start() throws Exception
     }
 
     @Override
-    public void close() throws IOException
-    {
+    public void close() throws IOException {
+        close(0);
+    }
+    
+    public void close(int timeout) throws IOException {
```
Comment (Randgalt) [inline curator-client/src/main/java/org/apache/curator/ConnectionState.java]: please rename to `timeoutMs`

File: curator-client/src/main/java/org/apache/curator/CuratorZookeeperClient.java [hunk 2]
```diff
@@ -213,18 +213,25 @@ public void start() throws Exception
...
+     */
+    public void close() {
+        close(0);
+    }
+            
     /**
      * Close the client
      */
-    public void close()
+    public void close(int timeout)
```
Comment (Randgalt) [inline curator-client/src/main/java/org/apache/curator/CuratorZookeeperClient.java]: Missing Javadoc for param `timeout` (also should be `timeoutMs`)

File: pom.xml [hunk 3]
```diff
@@ -59,7 +59,7 @@
         <jdk-version>1.7</jdk-version>
 
         <!-- versions -->
-        <zookeeper-version>3.5.3-beta</zookeeper-version>
+        <zookeeper-version>3.5.4-beta</zookeeper-version>
```
Comment (Randgalt) [inline pom.xml]: -1 one this. Let's leave dep/version changes for a separate PR

Comment (eolivelli) [reply pom.xml]: This is required. Zk#close(timeout) is only in 3.5.4.
I can log a separate JIRA and create a simple PR

Comment (Randgalt) [reply pom.xml]: No problem - just checking.

File: curator-client/src/main/java/org/apache/curator/CuratorZookeeperClient.java [hunk 4]
```diff
@@ -213,18 +213,29 @@ public void start() throws Exception
...
+    }
+            
+    /**
+     * Close this client object as the {@link #close() } method.
+     * This method will wait for internal resources to be released.
+     * 
+     * @param waitForShutdownTimeoutMs timeout (in milliseconds) to wait for resources to be released.
+     *                  Use zero or a negative value to skip the wait.
+     */
+    public void close(int waitForShutdownTimeoutMs)
```
Comment (eolivelli) [inline curator-client/src/main/java/org/apache/curator/CuratorZookeeperClient.java]: @Randgalt in internal ZK handle we are returning a boolean I think that in Curator this is not necessary.
I am leaving *void*

**Comment by eolivelli** [comment]
@Randgalt I think the patch is ready now

**Comment by eolivelli** [comment]
Kindly pinging @Randgalt

File: curator-client/src/main/java/org/apache/curator/HandleHolder.java [hunk 5]
```diff
@@ -73,15 +73,15 @@ String getNewConnectionString()
         return ((helperConnectionString != null) && !ensembleProvider.getConnectionString().equals(helperConnectionString)) ? helperConnectionString : null;
     }
 
-    void closeAndClear() throws Exception
+    void closeAndClear(int timeout) throws Exception
```
Comment (Randgalt) [inline curator-client/src/main/java/org/apache/curator/HandleHolder.java]: Please rename argument to `waitForShutdownTimeoutMs`

**Comment by Randgalt** [comment]
I'm +1 on this change. However, should we expose a top level method of setting the shutdown timeout? I suggest a new field in `CuratorFrameworkFactory.Builder`. Please add.

**Comment by eolivelli** [comment]
@Randgalt sure.
Added a defaultWaitForShutdownTimeoutMs property to the main builder.
Let me know if this sounds good to you, I hope I understood what you meant.

File: curator-framework/src/main/java/org/apache/curator/framework/CuratorFrameworkFactory.java [hunk 6]
```diff
@@ -401,6 +402,20 @@ public Builder zk34CompatibilityMode(boolean mode)
...
 
+        /**
+         * Set a default timeout for {@link CuratorZookeeperClient#close() }.
+         * The default is 0, which means that this feature is disabled.
+         *
+         * @since 4.0.2
+         * @param defaultWaitForShutdownTimeoutMs default timeout
+         * @return this
+         */
+        public Builder defaultWaitForShutdownTimeoutMs(int defaultWaitForShutdownTimeoutMs)
```
Comment (Randgalt) [inline curator-framework/src/main/java/org/apache/curator/framework/CuratorFrameworkFactory.java]: Shouldn't this be just `waitForShutdownTimeoutMs()`? Normally, `CuratorZookeeperClient#close()` isn't called by users directly. Setting this value will cause the shutdown timeout to always be used.

File: curator-client/src/main/java/org/apache/curator/HandleHolder.java (hunk 5 above)
Comment (eolivelli) [reply curator-client/src/main/java/org/apache/curator/HandleHolder.java]: done

File: curator-framework/src/main/java/org/apache/curator/framework/CuratorFrameworkFactory.java (hunk 6 above)
Comment (eolivelli) [reply curator-framework/src/main/java/org/apache/curator/framework/CuratorFrameworkFactory.java]: okay, done

**Comment by eolivelli** [comment]
@Randgalt renamed all the methods/properties

**Comment by Randgalt** [comment]
Damn - ZooKeeper 3.5.4-beta is Java 8, not Java 7. That's a major change for Curator. Let me see what I can do. I might be able to shoehorn this in via reflection.

**Comment by eolivelli** [comment]
I can do it if you prefer.
It is only a matter of calling that close(int) method of it exists

**Comment by Randgalt** [comment]
FYI  https://github.com/apache/curator/pull/269

**Comment by eolivelli** [comment]
I will rebase as you commit that change

**Comment by Randgalt** [comment]
> I will rebase as you commit that change

No need - I've already made the change locally. I also added a test. I'll merge both

**Comment by eolivelli** [comment]
Great ! Thank you @Randgalt

**Pull request merged**